example-utils/integration-test-utils/src/integration_test_utils.rs

```diff
@@ -45,41 +42,46 @@ pub async fn check_send_receive_message_discrepancy(
     }
 }
 
+// TODO: Fix this function to work with UUIDv7
 pub async fn check_messages_in_order(messages: Arc<Mutex<Vec<UMessage>>>) {
     let messages = messages.lock().await;
     if messages.is_empty() {
         return;
     }
 
-    // Step 1: Group messages by id.lsb
-    let mut grouped_messages: HashMap<u64, Vec<&UMessage>> = HashMap::new();
+    // Step 1: Group messages by source UUri
+    #[allow(clippy::mutable_key_type)]
+    let mut grouped_messages: HashMap<UUri, Vec<&UMessage>> = HashMap::new();
     for msg in messages.iter() {
-        let lsb = msg.attributes.as_ref().unwrap().id.lsb;
-        grouped_messages.entry(lsb).or_default().push(msg);
+        let source_uuri = msg
+            .attributes
+            .as_ref()
+            .unwrap()
+            .source
+            .as_ref()
+            .unwrap()
+            .clone();
+        grouped_messages.entry(source_uuri).or_default().push(msg);
     }
 
-    // Step 2: Check each group for strict increasing order of id.msb
-    for (lsb, group) in grouped_messages {
-        debug!("lsb: {lsb}");
+    // Step 2: Check each group for strict increasing order of id.msb first 48 bytes
+    for (source_uuri, group) in grouped_messages {
+        debug!("source_uuri: {source_uuri}");
         if let Some(first_msg) = group.first() {
-            let mut prev_msb = first_msg.attributes.as_ref().unwrap().id.msb;
+            let mut prev_timestamp = first_msg.attributes.as_ref().unwrap().id.msb >> 16;
             for msg in group.iter().skip(1) {
-                let curr_msb = msg.attributes.as_ref().unwrap().id.msb;
-                debug!("prev_msb: {prev_msb}, curr_msb: {curr_msb}");
-                if curr_msb <= prev_msb {
-                    panic!("!! -- Message ordering issue for lsb: {} -- !!", lsb);
+                let curr_timestamp = msg.attributes.as_ref().unwrap().id.msb >> 16;
```

Comment:
Im personally not a fan of shifts in code.
I usually like to go with a union struct, or bitvector. In rust tho, this gets really ugly
``` rust
#[repr(C)]
union MyUnion {
    value: u32,
    timestamp: Timestamps,
}

#[repr(C)]
#[derive(Copy, Clone)]
struct Timestamps {
    prev: u16,
    curr: u16,
}

fn main() {
    let u = MyUnion { value: 0xDEADBEEF };
    unsafe { 
        println!("0x{:X} 0x{:X}", u.timestamp.prev, u.timestamp.curr); 
    } // 0XDEAD BEEF 
}
```

Food for thought...

Reply:
Yes, Rust is like this haha

Struct ordering is not guaranteed in Rust unless we use repr(C), and then you get:

![image](https://github.com/user-attachments/assets/ed670ca3-a967-4ff9-9cd0-51b6b8e9e111)